openhsi/calibrate.py

```diff
@@ -164,22 +173,41 @@ def fit_HgAr_lines(self, top_k:int = 10,
 
             print(f"top_A_idx={top_A_idx}\nA[top_A_idx]={A[top_A_idx]}\nμ[top_A_idx]={μ[top_A_idx]}\nσ[top_A_idx]={σ[top_A_idx]}\nbrightest_peaks={brightest_peaks}")
 
-        #top_A_idx = np.flip(np.argsort(A))[:len(brightest_peaks)]
+        # interpolate with brightest spectral lines
         first_fit = np.poly1d( np.polyfit(np.sort(μ[top_A_idx]),brightest_peaks,1) )
         predicted_λ = first_fit(μ)
+        print(f"Predicted λ {predicted_λ} for column {μ}")
 
-        plt.plot(first_fit(np.arange(len(spectra))))
-        plt.xlabel("array index"); plt.ylabel("fitted wavelength")
+        plt.plot(μ[top_A_idx], brightest_peaks, "xr")
+        plt.plot(np.arange(len(spectra)), first_fit(np.arange(len(spectra))))
+        plt.legend(['Identified Peaks', 'Spectra']); plt.xlabel("array index"); plt.ylabel("digital number")
         plt.show()
 
-        closest_λ = np.array([ HgAr_lines[np.argmin(np.abs(HgAr_lines-λ))] for λ in predicted_λ])
-        top_A_idx = np.flip(np.argsort(A))[:max(min(top_k,len(HgAr_lines)),4)]
-        final_fit = np.poly1d( np.polyfit(μ[top_A_idx],closest_λ[top_A_idx] ,3) )
-        spec_wavelengths = final_fit(μ[top_A_idx])
+        closest_HgAr_line = np.empty(0)
+        matching_centroid = np.empty(0)
+        matching_A = np.empty(0)
+        for λ, mu, A in zip(predicted_λ, μ, A):
+            diff = np.min(np.abs(HgAr_lines - λ))
+            print(diff)
+            if diff < max_match_error:  # nm
+                closest_HgAr_line = np.append(
+                    closest_HgAr_line, HgAr_lines[np.argmin(np.abs(HgAr_lines - λ))]
```

Comment:
This could made neater using standard Python lists. I'm not so comfortable with how long this function is getting and it's taking me longer to understand what's going on.  
There's lots of diagnostics printing. enable using `verbose=True`? See my suggestion below
```python
closest_HgAr_line, matching_centroid, matching_A  = [], [], []
diffs = [np.abs(HgAr_lines-λ) for λ in predicted_λ]
for i in range(len(diffs)):
    if verbose: print(f"difference HgAr_lines - λ = {diffs[i]}")
    if diffs[i] < max_match_error: # nm
        closest_HgAr_line.append( HgAr_lines[np.argmin(np.abs(HgAr_lines - predicted_λ[i]))] )
        matching_centroid.append( mu[i] ); matching_A.append( A[i] )
```

Reply:
happy with something like that, I was not very happy with original version, just needed something working at the time.

Reply:
<img width="758" alt="image" src="https://user-images.githubusercontent.com/782659/150454502-eea7d85f-f770-40a9-a578-6b6fd70e3d13.png">

this does not seem to work as is.

Reply:
put I'm not sure my port from my code to current master was correct, let me check that.

Reply:
I forgot the `min` in 
`diffs = [np.abs(HgAr_lines-λ) for λ in predicted_λ]`
should be
`diffs = [np.min(np.abs(HgAr_lines-λ)) for λ in predicted_λ]`
?

Reply:
ahh, yup, that helps.

another issue is the indexing of lists with Numpy array 'top_A_idx'

could fix with np.asarray where matching_centroid and closest_HgAr_line are used, but it is a bit messy.

`final_fit = np.poly1d( np.polyfit(np.asarray(matching_centroid)[top_A_idx],np.asarray(closest_HgAr_line)[top_A_idx] ,3) )
 `

otherwise convert them to ndarray after building

`closest_HgAr_line= np.asarray(closest_HgAr_line)`
etc

or init as ndarray like my original one:
`closest_HgAr_line,matching_centroid,matching_A = [np.empty(0) for i in range(3)]`

Reply:
fixed by converting to ndarray at end.